### PR TITLE
JIT: use new phase objects more extensively

### PR DESCRIPTION
**Pull request by AndyAyersMS**
Convert the mainline compiler phases over so they are run via compiler or
action phases.

To keep jit dump size and checked jit throughput the same as it is now,
add phase whitelists to opt into pre phase dumps and checks triggered by the
phase objects on a per-phase basis. This allow us to gradually remove explicit
dumps and check invocations phase by phase, over time.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
Contributes to #2109. Will verify no diffs.

@dotnet/jit-contrib PTAL

File: src/coreclr/src/jit/phase.cpp [hunk 1]
```diff
@@ -27,8 +27,40 @@ void Phase::PrePhase()
...
 #ifdef DEBUG
+
+    // To help in the incremental conversion of jit activity to phases
+    // without greatly increasing dump size or checked jit time, we
+    // currently whitelist the phases that do pre-phase checks and
+    // dumps via the phase object, and not via explicit calls from
+    // the various methods in the phase.
+    //
+    // In the long run the aim is to get rid of all pre-phase checks
+    // and dumps, relying intstead on post-phase checks and dumps from
```
Comment (BruceForstall) [inline src/coreclr/src/jit/phase.cpp]: nit: typo: intstead

File: src/coreclr/src/jit/compiler.cpp [hunk 2]
```diff
@@ -4395,94 +4407,84 @@ void Compiler::compCompile(void** methodCodePtr, ULONG* methodCodeSize, JitFlags
...
 #endif // DEBUG
 
-#if defined(DEBUG) && defined(TARGET_XARCH)
-    if (opts.compStackCheckOnRet)
-    {
-        lvaReturnSpCheck                  = lvaGrabTempWithImplicitUse(false DEBUGARG("ReturnSpCheck"));
-        lvaTable[lvaReturnSpCheck].lvType = TYP_I_IMPL;
-    }
-#endif // defined(DEBUG) && defined(TARGET_XARCH)
+#if defined(DEBUG) && defined(_TARGET_XARCH_)
```
Comment (BruceForstall) [inline src/coreclr/src/jit/compiler.cpp]: Didn't you convert these #ifdefs to the old form? Shouldn't you leave them without leading/trailing underscores?

**Review by BruceForstall — CHANGES_REQUESTED** [review]
Seems like some of the larger `auto` clauses might be better extracted out as functions. But generally looks good to me (except the apparently incorrect `#ifdef`s.)

File: src/coreclr/src/jit/compiler.cpp (hunk 2 above)
Comment (AndyAyersMS) [reply src/coreclr/src/jit/compiler.cpp]: Thanks for spotting that.

This PR is a revival of some older work, so it has the old ifdef names. Need to keep reminding myself to watch for them.

**Review by BruceForstall — APPROVED** [review]

**Comment by AndyAyersMS** [comment]
@dotnet/jit-contrib anyone else want to look this over? I'm somewhat ambivalent about the lambdas, though we can always peel those out as named functions if they become too annoying.

**Pull request merged**